reframe/core/buildsystems.py

```diff
@@ -782,6 +792,94 @@ def generated_modules(self):
         return self._eb_modules
 
 
+class Spack(BuildSystem):
+    '''A build system for building test code using `Spack
+    <https://spack.io/>`__.
+
+    ReFrame will use Spack to build and install the code given an environment
+    and a set of specs.
+
+    .. versionadded:: 3.6.1
+
+    '''
+
+    #: The list of specs to build and install within the given environment.
+    #:
+    #: ReFrame will add the specs to the active environment by emititing the
+    #: following command:
+    #:
+    #: .. code-block:: bash
+    #:
+    #:    spack add spec1 spec2 spec3
+    #:
+    #: :type: :class:`List[str]`
+    #: :default: ``[]``
+    specs = fields.TypedField(typ.List[str])
+
+    #: Spack environment.
+    #: This field is required.
+    #:
+    #: ReFrame activates this environment to build and install the specified
+    #: packages:
+    #:
+    #: .. code-block:: bash
+    #:
+    #:    spack env activate -d <environment directory>
+    #:
+    #: :type: :class:`str` or :class:`None`
+    #: :default: :class:`None`
+    environment = fields.TypedField(str, type(None))
+
+    def __init__(self):
+        super().__init__()
+        self.specs = []
+        self.environment = None
+        self._prefix_save = None
+
+    def emit_build_commands(self, environ):
+        self._prefix_save = os.getcwd()
+        if not self.environment:
+            raise BuildSystemError(f"'environment' must not be empty")
+
+        ret = self._env_activate_cmds()
+        if self.specs:
+            specs_str = ' '.join(self.specs)
+            ret.append(f'spack add {specs_str}')
+
+        ret.append('spack install')
+        return ret
+
+    def _env_activate_cmds(self):
+        return [f'. $SPACK_ROOT/share/spack/setup-env.sh',
+                f'spack env activate -d {self.environment}']
+
+    def prepare_cmds(self):
+        cmds = self._env_activate_cmds()
+        if self.specs:
+            cmds.append('spack load ' + ' '.join(s for s in self.specs))
+
+        return cmds
+
+    def post_build(self, buildjob):
+        # Store the modules generated by Spack
+
+        with open(os.path.join(self._prefix_save, buildjob.stdout)) as fp:
+            out = fp.read()
+
+        matches = re.findall(r'\[\+\] \S+/spack/opt/spack/\S+/(\S+)',
+                             out)
+
+        self._modules = []
+        for mod in matches:
+            mod = os.path.basename(mod)
+            name, version, hashid = mod.split('-')
+            self._modules.append(f'{name}@{version}/{hashid}')
+
+    @property
+    def generated_modules(self):
+        return self._modules
```

Comment:
@victorusu @rsarm Do we really need this now? I have modified the code to automatically generate `spack load` for running. I'm asking because I'm getting a strange error when we are trying to retrieve the name, version and hash id above:

```
Traceback (most recent call last):
  File "/Users/karakasv/Repositories/reframe/reframe/frontend/executors/__init__.py", line 269, in _safe_call
    return fn(*args, **kwargs)
  File "/Users/karakasv/Repositories/reframe/reframe/core/pipeline.py", line 119, in _fn
    func(obj, *args, **kwargs)
  File "/Users/karakasv/Repositories/reframe/reframe/core/pipeline.py", line 133, in _wrapped
    return fn(*args, **kwargs)
  File "/Users/karakasv/Repositories/reframe/reframe/core/pipeline.py", line 1304, in compile_wait
    self.build_system.post_build(self._build_job)
  File "/Users/karakasv/Repositories/reframe/reframe/core/buildsystems.py", line 875, in post_build
    name, version, hashid = mod.split('-')
ValueError: too many values to unpack (expected 3)
```